src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractPullDiagnosticHandler.cs

```diff
@@ -129,6 +133,25 @@ private void OnDiagnosticsUpdated(object? sender, DiagnosticsUpdatedArgs updateA
             return progress.GetValues();
         }
 
+        private static bool IncludeDocument(Document document, string? clientName)
+        {
+            // Documents either belong to Razor or not.  We can determine this by checking if the doc has a span-mapping
+            // service or not.  If we're not in razor, we do not include razor docs.  If we are in razor, we only
+            // include razor docs.
+            var isRazorDoc = IsRazorDocument(document);
+            var wantsRazorDoc = clientName != null;
+
+            return wantsRazorDoc == isRazorDoc;
+        }
+
+        private static bool IsRazorDocument(Document document)
+        {
+            // Only razor docs have an ISpanMappingService, so we can use the presence of that to determine if this doc
+            // belongs to them.
+            var spanMapper = document.Services.GetService<ISpanMappingService>();
+            return spanMapper != null;
+        }
```

Comment:
Is there a reason this isn't using https://github.com/dotnet/roslyn/blob/80845e82045f1fc4d2e4e7eb489262403872bdbf/src/Workspaces/Core/Portable/Workspace/Host/DocumentService/DocumentPropertiesService.cs#L20?

Reply:
@dibarbet ?

Reply:
I think we could use either.  I guess the DiagnosticsLspClientName is more appropriate since this is diagnostics?

Reply:
Do we need to keep DiagnosticsLspClientName after this effort though? I thought perhaps we could remove DocumentPropertiesService entirely in fact

Reply:
i'm not really understanding what DiagnosticsLspClientName is used for.  It's also unclear what i would be checking for at this level.  Just it being non-null?  seems a bit strange IMO.

Reply:
> Do we need to keep DiagnosticsLspClientName after this effort though? I thought perhaps we could remove DocumentPropertiesService entirely in fact

I can def see if that's possible.